### PR TITLE
Add nginx config and update dockerfile to consume it

### PR DESCRIPTION
**Pull request by imnotteixeira**
Currently, when manually going to `mainURL`/something will give a 404 nginx error if the page does not exist, however, with this PR, it will redirect to index.html, so that react-router can deal with the route inside it

### CODE REVIEW

File: Dockerfile-prod [hunk 1]
```diff
@@ -24,12 +24,10 @@ COPY .env* ./
 # Building the image
 RUN npm run build
 
-# Copy env files
```
Comment (miguelpduarte) [inline Dockerfile-prod]: Why was this removed?

Comment (miguelpduarte) [reply Dockerfile-prod]: Hmm, this seems to be the old wrongly placed one, but was this not already removed?

File: Dockerfile-prod [hunk 2]
```diff
@@ -24,12 +24,10 @@ COPY .env* ./
...
-COPY .env* ./
-
 # Then, use nginx to serve the built files
 # See https://hub.docker.com/_/nginx
 FROM nginx:alpine
 
 # Copying the built files into the nginx image, to the default served directory
 COPY --from=build /usr/src/app/build /usr/share/nginx/html
+COPY ./nginx-config/nijobs.conf /etc/nginx/conf.d/default.conf
```
Comment (miguelpduarte) [inline Dockerfile-prod]: Why have a directory? Also, the `./` in the copy from looks weird, and is unnecessary.
The reason would be to override all files in the directory, but that only applies using volumes so it would be the same to `RUN rm /etc/nginx/conf.d/default.conf` and then `COPY nijobs-nginx.conf /etc/nginx/conf.d/whatever.conf`

File: Dockerfile-prod (hunk 1 above)
Comment (imnotteixeira) [reply Dockerfile-prod]: I did a git pull on develop before this, so it must have slipped, it's no big deal actually

File: Dockerfile-prod (hunk 2 above)
Comment (miguelpduarte) [inline Dockerfile-prod]: Why have a directory? Also, the `./` in the copy from looks weird, and is unnecessary.
The reason I see would be to override all files in the directory, but that only applies using volumes so it would be the same to just `RUN rm /etc/nginx/conf.d/default.conf` and then `COPY nijobs-nginx.conf /etc/nginx/conf.d/whatever.conf`

File: nginx-config/nijobs.conf [hunk 3]
```diff
@@ -0,0 +1,37 @@
+server {
+        # listen on port 80
+        listen 80;
+        # save logs here
+        access_log /var/log/nginx/access.log main;
+
+        # where the root here
+        root /usr/share/nginx/html;
```
Comment (miguelpduarte) [inline nginx-config/nijobs.conf]: This can be excluded, as it is the default, I believe

File: nginx-config/nijobs.conf [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        location ~* \.(?:jpg|jpeg|gif|png|ico|cur|gz|svg|svgz|mp4|ogg|ogv|webm|htc)$ {
+          expires 1M;
+          access_log off;
+          add_header Cache-Control "public";
+        }
+
+        # Javascript and CSS files
+        location ~* \.(?:css|js)$ {
+            try_files $uri =404;
+            expires 1y;
```
Comment (miguelpduarte) [inline nginx-config/nijobs.conf]: Is this desired? Will new versions of files hit cache and thus not update?

File: nginx-config/nijobs.conf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            # as directory, then fall back to redirecting to index.html
+            try_files $uri $uri/ /index.html;
+        }
+
+        # Media: images, icons, video, audio, HTC
+        location ~* \.(?:jpg|jpeg|gif|png|ico|cur|gz|svg|svgz|mp4|ogg|ogv|webm|htc)$ {
+          expires 1M;
+          access_log off;
+          add_header Cache-Control "public";
+        }
```
Comment (miguelpduarte) [inline nginx-config/nijobs.conf]: Is this necessary? Can't we let nginx/the browser decide defaults?

File: nginx-config/nijobs.conf [hunk 6]
```diff
@@ -0,0 +1,37 @@
+server {
+        # listen on port 80
+        listen 80;
+        # save logs here
+        access_log /var/log/nginx/access.log main;
+
+        # where the root here
+        root /usr/share/nginx/html;
+        # what file to server as index
```
Comment (miguelpduarte) [inline nginx-config/nijobs.conf]: typo

File: nginx-config/nijobs.conf [hunk 7]
```diff
@@ -0,0 +1,37 @@
+server {
+        # listen on port 80
+        listen 80;
+        # save logs here
+        access_log /var/log/nginx/access.log main;
+
+        # where the root here
```
Comment (miguelpduarte) [inline nginx-config/nijobs.conf]: typo?

File: nginx-config/nijobs.conf (hunk 3 above)
Comment (imnotteixeira) [reply nginx-config/nijobs.conf]: From what I've seen, a lot of examples had /var/www as default, we should keep it in order to make it clearer

Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: The fact that the Dockerfile worked without an nginx.conf disproves that. The default for nginx's static file serving is `/usr/share/nginx/html`

File: nginx-config/nijobs.conf (hunk 4 above)
Comment (imnotteixeira) [reply nginx-config/nijobs.conf]: I might be wrong, but in this setup, if there are "new files" there will be another container generated, therefore another nginx instance and no cache problems

Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: This cache is in the browsers. This is the header setting for that expiration, if I'm not mistaken

Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: It would make no sense for nginx to serve outdated versions of local files 🙃 the overhead would be bigger on creating that cache system than just serving them outright

File: Dockerfile-prod (hunk 2 above)
Comment (imnotteixeira) [reply Dockerfile-prod]: It is not necessary to have a rm + cp. Instead we just cp to the default, which will override and work, since it's in an isolated environment, it won't cause a problem when overriding.

File: nginx-config/nijobs.conf (hunk 3 above)
Comment (imnotteixeira) [reply nginx-config/nijobs.conf]: That might be true, however, it doesn't hurt to have it there to make it clearer for future devs.

File: nginx-config/nijobs.conf (hunk 4 above)
Comment (imnotteixeira) [reply nginx-config/nijobs.conf]: I've been researching, and this is actually useful. Every time we build and there is a change in the files, they change their name (they are appended with some kind of hash I believe). Therefore, if a build only changes some files that might only update some built ones and keep the others in cache, which is good

File: Dockerfile-prod (hunk 2 above)
Comment (miguelpduarte) [reply Dockerfile-prod]: Sure. However, using a directory just for holding an nginx conf file seems weird to me

File: nginx-config/nijobs.conf (hunk 3 above)
Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: Sure, I'm fine with having it this way

File: nginx-config/nijobs.conf (hunk 5 above)
Comment (imnotteixeira) [reply nginx-config/nijobs.conf]: This ensures consistency and a good experience for everyone. I don't know the default expiry date for images, but considering that the only images we'll have are either static (our logo, some helper svgs, ...) or almost static (companies profile logos - and even these will have an appended id, making sure that if they are updated, they change) I think that 1Month is a good value.

Regarding the Cache-Control: public - it's actually better since it will allow us to cache these images even with https, if it were private, which is the default, the images wouldn't be cached

This is a really good doc on web caches: https://www.mnot.net/cache_docs/

File: nginx-config/nijobs.conf (hunk 4 above)
Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: Okay, good to go then!

File: nginx-config/nijobs.conf (hunk 5 above)
Comment (miguelpduarte) [reply nginx-config/nijobs.conf]: Sounds good then!
